### PR TITLE
release generator random figure

### PR DESCRIPTION
**Pull request by u3IOm4uk**

### CODE REVIEW

File: src/main/java/core/basesyntax/Figure.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package core.basesyntax;
+
+import core.basesyntax.enums.Color;
+import core.basesyntax.enums.FigureName;
+import core.basesyntax.interfaces.Area;
+import core.basesyntax.interfaces.Draw;
+
+public abstract class Figure implements Area, Draw {
+    private Color color = new ColorSupplier().getRandomColor();
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Figure.java:9]:
```suggestion
    private Color color;
```

Comment (kozhukhovsky) [reply src/main/java/core/basesyntax/Figure.java:9]: It is necessary not to set the color here, but to pass it through the constructor during creation.

File: src/main/java/core/basesyntax/Main.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+                figureArr[i] = figureSupplier.getRandomFigure();
+                System.out.println(figureArr[i].draw());
+            }
+            if (i >= figureArr.length / 2) {
+                figureArr[i] = figureSupplier.getDefaultFigure();
+                System.out.println(figureArr[i].draw());
+            }
+        }
+    }
+
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Main.java:21]:
```suggestion
```

File: src/main/java/core/basesyntax/enums/Color.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.enums;
+
+public enum Color {
+    black,
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/enums/Color.java:4]:
```suggestion
    BLACK,
```

Comment (kozhukhovsky) [reply src/main/java/core/basesyntax/enums/Color.java:4]: in enam all items come in the upper case

File: src/main/java/core/basesyntax/enums/FigureName.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.enums;
+
+public enum FigureName {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/enums/FigureName.java:3]: same

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            case Square:
+                double side = random.nextInt(boundMin, boundMax);
+                return new Square(side);
+            default:
+                return null;
+        }
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle(10, Color.white);
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:42]: move 10 to the constant

File: src/main/java/core/basesyntax/interfaces/Area.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.interfaces;
+
+public interface Area {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/interfaces/Area.java:3]:
```suggestion
public interface AreaCalculator {
```

File: src/main/java/core/basesyntax/interfaces/Draw.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.interfaces;
+
+public interface Draw {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/interfaces/Draw.java:3]:
```suggestion
public interface Drawable {
```

**Review by kozhukhovsky — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import core.basesyntax.enums.Color;
+import java.util.Random;
+
+public class ColorSupplier {
+    private int indexColor;
+    private final Random random = new Random();
+
+    public Color getRandomColor() {
+        indexColor = random.nextInt(Color.values().length);
+        return Color.values()[indexColor];
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/ColorSupplier.java:12]: Return color in String value

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package core.basesyntax;
+
+import core.basesyntax.enums.Color;
+import java.util.Random;
+
+public class ColorSupplier {
+    private int indexColor;
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/ColorSupplier.java:7]: This field is redundant

File: src/main/java/core/basesyntax/Figure.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        return color;
+    }
+
+    public void setColor(Color color) {
+        this.color = color;
+    }
+
+    public void setColor() {
+        this.color = new ColorSupplier().getRandomColor();
+    }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/Figure.java:30]: Don`t init color in this class

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        this.radius = radius;
+        this.setFigureName(FigureName.CIRCLE);
+        this.setColor();
+    }
+
+    public Circle(double radius, Color color) {
+        this.radius = radius;
+        this.setFigureName(FigureName.CIRCLE);
+        this.setColor(color);
+    }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/figure/Circle.java:20]: You don`t need 2 constructors here

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            case RECTANGLE:
+                double sideBasis = random.nextInt(boundMin, boundMax);
+                height = random.nextInt(boundMin, boundMax);
+                return new Rectangle(sideBasis, height);
+            case RIGHTTRIANGLE:
+                double firstLeg = random.nextInt(boundMin, boundMax);
+                double secondLeg = random.nextInt(boundMin, boundMax);
+                return new RightTriangle(firstLeg, secondLeg);
+            case SQUARE:
+                double side = random.nextInt(boundMin, boundMax);
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:35]: Duplication code, you can take out in separate method

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Figure.java (hunk 10 above)
Comment (Ivan95kos) [reply src/main/java/core/basesyntax/Figure.java:29]: not fixed

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Random;
+
+public class FigureSupplier {
+    private final int boundMax = 20;
+    private final int boundMin = 1;
+    private final int defaultRadius = 10;
+    private final Random random = new Random();
+    private int indexFigureName;
+
+    public int randomParameters() {
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:14]: Methods that use only in class should be private
```suggestion
    private int randomParameters() {
```

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
+package core.basesyntax.figure;
+
+import core.basesyntax.Figure;
+import core.basesyntax.enums.FigureName;
+import java.util.Random;
+
+public class FigureSupplier {
+    private final int boundMax = 20;
+    private final int boundMin = 1;
+    private final int defaultRadius = 10;
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:10]: Make them constants

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    private final Random random = new Random();
+    private int indexFigureName;
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:12]: This field should be in the method

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+                return new Rectangle(sideBasis, height);
+            case RIGHTTRIANGLE:
+                double firstLeg = randomParameters();
+                double secondLeg = randomParameters();
+                return new RightTriangle(firstLeg, secondLeg);
+            case SQUARE:
+                double side = randomParameters();
+                return new Square(side);
+            default:
+                return null;
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:41]:
```suggestion
                return getDefaultFigure();
```

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        }
+    }
+
+    public Figure getDefaultFigure() {
+        Figure circle = new Circle(defaultRadius);
+        circle.setColor("WHITE");
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:47]: Use Enum and constant
```suggestion
        circle.setColor(DEFAULT_COLOR);
```

File: src/main/java/core/basesyntax/interfaces/Drawable.java [hunk 18]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.interfaces;
+
+public interface Drawable {
+    String draw();
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/interfaces/Drawable.java:4]: You should draw the Figure inside of method
```suggestion
    void draw();
```

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
Please, fix all previous comments

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+public class ColorSupplier {
+    private final Random random = new Random();
+
+    public String getRandomColor() {
+        return Color.values()[random.nextInt(Color.values().length)].toString();
+    }
+
+    public String getDefaultColor() {
+        return "WHITE";
+    }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ColorSupplier.java:15]: Make a constant `DEFAULT_COLOR` in `FigureSupplier`
```suggestion
```

File: src/main/java/core/basesyntax/figure/FigureSupplier.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+                double side = randomParameters();
+                color = colorSupplier.getRandomColor();
+                return new Square(side, color);
+            default:
+                return getDefaultFigure();
+        }
+    }
+
+    public Figure getDefaultFigure() {
+        final int defaultRadius = 10;
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/figure/FigureSupplier.java:52]: it must be a constant
```suggestion
```

**Review by Ivan95kos — APPROVED** [review]
Great job, with minor comments

**Pull request closed without merging**